ArduCopter/mode_guided.cpp

```diff
@@ -768,6 +779,163 @@ bool ModeGuided::limit_check()
     return false;
 }
 
+//loiter turns(circle) mode
+
+void ModeGuided::do_circle(const mavlink_command_long_t &packet)
+{
+    const Location circle_center = loc_from_param(packet);
+
+    // calculate radius
+    uint8_t circle_radius_m = radius_from_param(packet); // circle radius held in high byte of p1
+
+    //store the number of turns
+    turn_count = (uint8_t)packet.param1;
+
+    // move to edge of circle (verify_circle) will ensure we begin circling once we reach the edge
+    circle_movetoedge_start(circle_center, circle_radius_m);
+}
+
+Location ModeGuided::loc_from_param(const mavlink_command_long_t &packet)
+{    
+    Location ret((int32_t)packet.param5,(int32_t)packet.param6,(int32_t)packet.param7,copter.current_loc.get_alt_frame());
+
+    // use current lat, lon if zero
+    if (ret.lat == 0 && ret.lng == 0) {
+        ret.lat = copter.current_loc.lat;
+        ret.lng = copter.current_loc.lng;
+    }
+    // use current altitude if not provided
+    if (ret.alt == 0) {
+        // set to current altitude but in command's alt frame
+        int32_t curr_alt;
+        if (copter.current_loc.get_alt_cm(ret.get_alt_frame(),curr_alt)) {
+            ret.set_alt_cm(curr_alt, ret.get_alt_frame());
+        } else {
+            // default to current altitude as alt-above-home
+            ret.set_alt_cm(copter.current_loc.alt,
+                           copter.current_loc.get_alt_frame());
+        }
+    }
+    return ret;
+
+}
+
+uint8_t ModeGuided::radius_from_param(const mavlink_command_long_t &packet)
+{
+    if ((uint8_t)packet.param3 == 0) {
+        return (uint8_t)(copter.circle_nav->get_radius()/100);
+    }
+    return (uint8_t)packet.param3;
+}
+
+void ModeGuided::circle_movetoedge_start(const Location &circle_center,float radius_m)
+{
+    // set circle center
+    copter.circle_nav->set_center(circle_center);
+
+    // set circle radius
+    if (!is_zero(radius_m)) {
+        copter.circle_nav->set_radius(radius_m * 100.0f);
+    }
+
+    // check our distance from edge of circle
+    Vector3f circle_edge_neu;
+    copter.circle_nav->get_closest_point_on_circle(circle_edge_neu);
+    float dist_to_edge = (inertial_nav.get_position() - circle_edge_neu).length();
+
+    // if more than 3m then fly to edge
+    if (dist_to_edge > 300.0f) {
+        // set the state to move to the edge of the circle
+        guided_mode = Guided_CircleMoveToEdge;
```

Comment:
at the exception of this, circle_movetoedge_start() is the same as the AUTO one's, we should factorize

Reply:
sir, the circle_movetoedge_start() and circle_run() have differences from the AUTO one's.A similar case is made in the use_pilot_yaw() function which is present in : 
- [Mode Auto](https://github.com/ArduPilot/ardupilot/blob/master/ArduCopter/mode.h#L414)
- [Mode Guided](https://github.com/ArduPilot/ardupilot/blob/master/ArduCopter/mode.h#L855)
- [Mode RTL](https://github.com/ArduPilot/ardupilot/blob/master/ArduCopter/mode.h#L1141)
I was thinking about adding new parameters in these functions in the Mode class that can help to differentiate between functions. I am a little confused about refactoring these parts.Would like to know further directions I should take.
Thanks, @khancyr .